### DISCUSSION TITLE
Unable to add Action.Arguments to task defenition

### DISCUSSION THREAD

**Discussion by ghost** [dahall/TaskScheduler#911] [Q&A] [answered]
I'm working on a C# console/winforms desktop app using .NET Framework 4.7.2.  It monitors some local drive properties and displays a status message. Status message display is executed via Task Scheduler with the task programmatically defined and registered (using your very much appreciated NuGet package).  XmlWriter/StringBuilder generates Trigger and Action code.  I’ve got everything working except Action Arguments.

==> This bit works
```
w.WriteStartElement("Actions");
w.WriteStartElement("Exec");
w.WriteElementString("Command", $"\"{startLocation}\"");
w.WriteEndElement();
w.WriteEndElement();
```

This code with a Command peer Element added for Arguments
```
w.WriteStartElement("Actions");
w.WriteStartElement("Exec");
w.WriteElementString("Command", $"\"{startLocation}\"");
w.WriteEndElement();
w.WriteElementString("Arguments", $"\"{xName}\"");
w.WriteEndElement();
w.WriteEndElement();
w.WriteEndElement();
```
Throws a runtime Exception: Token EndElement in state EndRootElement would result in an invalid XML document. 

And with this code
```
w.WriteStartElement("Actions");
w.WriteStartElement("Exec");
w.WriteElementString("Command", $"\"{startLocation}\", \"{xName}\"");
w.WriteEndElement();
w.WriteEndElement();
```
The task is created but won't run.

I've also tried:
```
ExecAction ea = new ExecAction();
ea.Path = $"\"{startLocation}\"";
ea.Arguments = $"{xName}";
td.Actions.Add(ea); 
```
But it doesn't get registered; just seems to be ignored.  I also tried commenting out the XML  Action block in an attempt to use the td.Actions approach but that errored out on "foreach (var line in GETXML) td.XmlText = $"{line.XMLcode}";".  After uncommenting the XML Action block the code (without Arguments) again ran fine.

Any assistance will be much appreciated. 

**_EDIT:_**
Found the problem.  For other novices “WriteElementString” does not need a “WriteEndElement” so working code  is:
```
w.WriteStartElement("Actions");
w.WriteStartElement("Exec");
w.WriteElementString("Command", $"\"{startLocation}\"");
w.WriteElementString("Arguments", $"\"{xName}\"");
w.WriteEndElement(); // close Exec
w.WriteEndElement(); // close Actions

```I have the setting “CloseOutput” set to true in case that’s salient.

However, as dahall states below working directly with the properties is much preferred to mucking about with XML if at all possible.

**Comment by dahall** [comment]
@ArtHansen Why are you trying to manipulate the XML? You can, of course, create an XML stream and then register it using `TaskFolder.RegisterTask` or you can set that stream to the `TaskDefinition.XmlText` property and then register the `TaskDefinition` using `TaskFolder.RegisterTaskDefinition`. However, I have found that to be more error prone than just using the properties of the different classes in the library as in your last code snippet with `ExecAction`. I would go that direction and then make sure that after any changes you re-register the task. (See example on https://github.com/dahall/TaskScheduler/wiki/Examples#get-and-delete-a-task and use your `ExecAction` code instead of the trigger code as shown).

**Reply by ghost** [reply]
Thank you for the prompt response.  I’m using XML because I couldn’t figure out how to set the MonthsOfYear (see below) property to a variable (variables are set via a user input form):
```
int[] dates = new int[_dates.Count - 1];
for (int i = 0; i < _dates.Count; i++)
{
	foreach (var d in _dates)
	{
		dates.SetValue(d, i);
	}
}
List<string> bldMon = new List<string>();
foreach (var mo in _months)
{
	bldMon.Add($"MonthsOfYear.{mo}");
}
string months = string.Join(" | ", bldMon);
DateTime.TryParse(TimeDTP.Text, out DateTime StartMe);

MonthlyTrigger mt = new MonthlyTrigger();
mt.StartBoundary = StartMe;
mt.DaysOfMonth = dates;
mt.MonthsOfYear = (MonthsOfTheYear)months;
```

Error CS0029 Cannot (implicitly without cast) convert type 'string' to 'Microsoft.Win32.TaskScheduler.MonthsOfTheYear'

Since I never got it to compile I don’t know if StartBoundary and DaysOfMonth would run successfully with the variables.  I would much prefer the properties route if I could incorporate variables into the property values.

**Comment by dahall** [comment]
To get lists of those enum values and their corresponding names, I have done the following:
```cs
// Get MonthsOfTheYear values and current culture names for each month
var monthValues = Enum.GetValues(typeof(MonthsOfTheYear)).Cast<MonthsOfTheYear>().Take(12).ToArray();
var monthNames = monthValues.Select(v => TaskEnumGlobalizer.GetString(v)).ToArray();
```

`StartBoundary` is just a `DateTime` value.

**Comment by dahall** [comment, marked as answer]
Here's a little snippet to demonstrate a little more elegant solution:
```cs
public class Element<T>
{
    public Element(T value, string text) { Value = value; Text = text; }
    public T Value { get; set; }
    public string Text { get; set; }
    public bool Selected { get; set; }
    public override string ToString() { return Text ?? ""; }
}

public static void Main()
{
    // Build a list of elements with the enum values and their localized text values
    List<Element<MonthsOfTheYear>> months = Enum.GetValues(typeof(MonthsOfTheYear)).Cast<MonthsOfTheYear>().Take(12).Select(v => new Element<MonthsOfTheYear>(v, TaskEnumGlobalizer.GetString(v))).ToList();
    
    // TODO: Use 'months' to populate UI and set the 'Element<T>.Selected' property, like following
    months[0].Selected = true;
    months[6].Selected = true;
    
    // Get the accumulated value of all selected elements
    MonthsOfTheYear moty = months.Where(m => m.Selected).Aggregate((MonthsOfTheYear)0, (r, m) => r |= m.Value);
}
```

**Reply by ghost** [reply]
Thank you very much.  As you can tell I'm a C# novice and this is my first exposure to Enum.  Looks quite useful; I'll attempt to incorporate your "snippet" into my logic ...

**_EDIT:_**
That's great.  I don't yet fully understand all your code magic but I at least understand it enough to make work. (and to leverage it)   One interesting item: if I edit the task in Scheduler I get this message:
![image](https://user-images.githubusercontent.com/91940316/136665830-50fe2d94-4cd7-4099-988c-7f18d7723e8a.png)

It must be an esoteric exception since there is no "Cancel" button.   I set Compatibility to V2_3 (Win10) but didn't solve the issue.

For other novices the working code (which replaces about 200: lines of XML based code)
```
private void MonthDates(string xName)
{
	string taskName = $"_BURSprop_{xName}";
	TaskDefinition td = TaskService.Instance.NewTask();

	td.RegistrationInfo.Description = "Display Backup Readiness Status";

	td.Principal.LogonType = TaskLogonType.InteractiveToken;
	td.Principal.RunLevel = TaskRunLevel.Highest;

	DateTime.TryParse(TimeDTP.Text, out DateTime StartMe);
	DateTime EndMe = StartMe.AddYears(100);

	int[] dates = new int[_dates.Count];
	for (int i = 0; i < _dates.Count; i++)
	{
		dates.SetValue(_dates[i], i);
	}

	// set the 'Element<T>.Selected' property
	List<string> months = new List<string>();
	foreach (CheckBox chk in MonthlyPanel_2.Controls.OfType<CheckBox>())
	{
		if (chk.Checked && chk.Name.Contains("_mon"))
		{ months.Add(chk.Text); }
	}
	for (int i = 0; i < _Months.Count; i++)
	{
		string mo = _Months[i].ToString();
		if (months.Contains(mo)) _Months[i].Selected = true;
	}
	MonthsOfTheYear moty = _Months.Where(m => m.Selected).Aggregate(
		(MonthsOfTheYear)0, (r, m) => r |= m.Value);

	MonthlyTrigger mt = new MonthlyTrigger();
	mt.StartBoundary = StartMe;
	mt.EndBoundary = EndMe;
	mt.DaysOfMonth = dates;
	mt.MonthsOfYear = moty;
	td.Triggers.Add(mt);

	string startLocation = exeName();
	ExecAction ea = new ExecAction();
	ea.Path = $"\"{startLocation}\"";
	ea.Arguments = $"\"{xName}\"";
	td.Actions.Add(ea);

	td.Settings.AllowDemandStart = true;
	td.Settings.StartWhenAvailable = true;
	td.Settings.Compatibility = TaskCompatibility.V2_3;

	TaskService.Instance.RootFolder.RegisterTaskDefinition(taskName, td);
}

```

**Comment by dahall** [comment]
Also, consider using the [`TriggerEditDialog`](https://github.com/dahall/TaskScheduler/wiki/TriggerSamples#showing-the-editor-for-a-trigger) rather than rolling your own UI.

**Reply by ghost** [reply]
I tried that but (for me) it really broke the aesthetic flow of the app ...
